File: app/src/main/java/me/devsaki/hentoid/workers/UpdateDownloadWorker.java
package me.devsaki.hentoid.workers;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.WorkerParameters;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.notification.update.UpdateFailedNotification;
import me.devsaki.hentoid.notification.update.UpdateInstallNotification;
import me.devsaki.hentoid.notification.update.UpdateProgressNotification;
import me.devsaki.hentoid.util.file.FileHelper;
import me.devsaki.hentoid.util.network.HttpHelper;
import me.devsaki.hentoid.util.notification.BaseNotification;
import me.devsaki.hentoid.workers.data.UpdateDownloadData;
import okhttp3.Response;
import okhttp3.ResponseBody;
import timber.log.Timber;


/**
 * Worker responsible for downloading app udate (APK)
 */
public class UpdateDownloadWorker extends BaseWorker {

    public UpdateDownloadWorker(
            @NonNull Context context,
            @NonNull WorkerParameters parameters) {
        super(context, parameters, R.id.update_download_service, null);
    }

    public static boolean isRunning(@NonNull Context context) {
        return isRunning(context, R.id.update_download_service);
    }

    @Override
    protected BaseNotification getStartNotification() {
        return new UpdateProgressNotification();
    }

    @Override
    protected void onInterrupt() {
        // Nothing
    }

    @Override
    protected void onClear() {
        // Nothing
    }

    @Override
    protected void getToWork(@NonNull Data input) {
        UpdateDownloadData.Parser data = new UpdateDownloadData.Parser(getInputData());
        String apkUrl = data.getUrl();

        try {
            downloadUpdate(apkUrl);
        } catch (IOException e) {
            Timber.w(e, "Update download failed");
            notificationManager.notifyLast(new UpdateFailedNotification(apkUrl));
        }
    }

    private void downloadUpdate(String apkUrl) throws IOException {
        Context context = getApplicationContext();
        Timber.w(context.getResources().getString(R.string.starting_download));

        File file = new File(context.getExternalCacheDir(), "hentoid.apk");
        file.createNewFile();

        Response response = HttpHelper.getOnlineResource(apkUrl, null, false, false, false);
        Timber.d("DOWNLOADING APK - RESPONSE %s", response.code());
        if (response.code() >= 300) throw new IOException("Network error " + response.code());

        ResponseBody body = response.body();
        if (null == body)
            throw new IOException("Could not read response : empty body for " + apkUrl);

        long size = body.contentLength();
        if (size < 1) size = 1;

        Timber.d("WRITING DOWNLOADED APK TO %s (size %.2f KB)", file.getAbsolutePath(), size / 1024.0);
        byte[] buffer = new byte[FileHelper.FILE_IO_BUFFER_SIZE];
        int len;
        long processed = 0;
        int iteration = 0;
        try (InputStream in = body.byteStream(); OutputStream out = FileHelper.getOutputStream(file)) {
            while ((len = in.read(buffer)) > -1) {
                processed += len;
                if (0 == ++iteration % 50) // Notify every 200KB
                    updateNotificationProgress(Math.round(processed * 100f / size));
                out.write(buffer, 0, len);
            }
            out.flush();
        }
        Timber.d("Download successful");

        notificationManager.notifyLast(new UpdateInstallNotification(FileHelper.getFileUriCompat(context, file)));
    }

    private void updateNotificationProgress(int processPc) {
        Timber.v("Download progress: %s%%", processPc);
        notificationManager.notify(new UpdateProgressNotification(processPc));
    }
}

File: app/src/main/java/me/devsaki/hentoid/workers/UpdateDownloadWorker.kt
package me.devsaki.hentoid.workers

import android.content.Context
import android.net.Uri
import android.webkit.MimeTypeMap
import androidx.documentfile.provider.DocumentFile
import androidx.work.Data
import androidx.work.WorkerParameters
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import me.devsaki.hentoid.R
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.events.CommunicationEvent
import me.devsaki.hentoid.events.ProcessEvent
import me.devsaki.hentoid.notification.appUpdate.UpdateFailedNotification
import me.devsaki.hentoid.notification.appUpdate.UpdateInstallNotification
import me.devsaki.hentoid.notification.appUpdate.UpdateProgressNotification
import me.devsaki.hentoid.util.download.downloadToFile
import me.devsaki.hentoid.util.file.getFileUriCompat
import me.devsaki.hentoid.util.file.legacyFileFromUri
import me.devsaki.hentoid.util.notification.BaseNotification
import me.devsaki.hentoid.workers.data.UpdateDownloadData
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.roundToInt

val APK_MIMETYPE = MimeTypeMap.getSingleton().getMimeTypeFromExtension("apk")

class UpdateDownloadWorker(context: Context, parameters: WorkerParameters) :
    BaseWorker(context, parameters, R.id.update_download_service, null) {

    private var progressPc = 0f
    private val killSwitch = AtomicBoolean(false)

    companion object {
        fun isRunning(context: Context): Boolean {
            return isRunning(context, R.id.update_download_service)
        }
    }

    init {
        EventBus.getDefault().register(this)
    }

    override fun getStartNotification(): BaseNotification {
        return UpdateProgressNotification()
    }

    override fun onInterrupt() {
        // Nothing
    }

    override suspend fun onClear(logFile: DocumentFile?) {
        EventBus.getDefault().unregister(this)
    }

    override suspend fun getToWork(input: Data) {
        val data = UpdateDownloadData.Parser(inputData)
        val apkUrl = data.url

        try {
            withContext(Dispatchers.IO) {
                downloadUpdate(apkUrl)
            }
        } catch (e: IOException) {
            Timber.w(e, "Update download failed")
            notificationManager.notifyLast(UpdateFailedNotification(apkUrl))
        }
    }

    @Throws(IOException::class)
    private suspend fun downloadUpdate(apkUrl: String) {
        Timber.d("DOWNLOADING APK")
        EventBus.getDefault().post(
            ProcessEvent(ProcessEvent.Type.PROGRESS, R.id.update_download_service)
        )
        val apk = downloadToFile(
            applicationContext,
            Site.NONE,
            apkUrl,
            emptyList(),
            Uri.fromFile(applicationContext.externalCacheDir),
            "hentoid.apk",
            isCanceled = { killSwitch.get() },
            resourceId = 0,
            forceMimeType = APK_MIMETYPE
        ) {
            progressPc = it
            if (0 == (progressPc.roundToInt() % 5)) launchProgressNotification()
        }

        if (killSwitch.get()) {
            notificationManager.cancel()
            return
        }

        apk?.let {
            Timber.d("Download successful")
            legacyFileFromUri(it)?.let { file ->
                // Must use getFileUriCompat to avoid being molested by Android
                val uri = getFileUriCompat(applicationContext, file)
                EventBus.getDefault().post(
                    CommunicationEvent(
                        CommunicationEvent.Type.APK_AVAILABLE,
                        message = uri.toString()
                    )
                )
                notificationManager.notifyLast(
                    UpdateInstallNotification(uri)
                )
            }
        } ?: run {
            Timber.d("Download failed")
        }
    }

    @Subscribe
    fun onCommunicationEvent(event: CommunicationEvent) {
        if (event.recipient != CommunicationEvent.Recipient.UPDATE_WORKER) return
        if (event.type == CommunicationEvent.Type.CANCEL) killSwitch.set(true)
    }

    override fun runProgressNotification() {
        Timber.v("Download progress: %s%%", progressPc.roundToInt())
        EventBus.getDefault().post(
            ProcessEvent(
                ProcessEvent.Type.PROGRESS,
                R.id.update_download_service,
                0,
                0, 0, progressPc
            )
        )
        notificationManager.notify(UpdateProgressNotification(progressPc.roundToInt()))
    }
}